pysmt/solvers/msat.py

```diff
@@ -574,3 +579,81 @@ def declare_variable(self, var):
                                                  tp)
             self.symbol_to_decl[var] = decl
             self.decl_to_symbol[mathsat.msat_decl_id(decl)] = var
+
+
+# Check if we are working on a patched version of the mathsat API
+if hasattr(mathsat, "MSAT_EXIST_ELIM_ALLSMT_FM"):
+    class MSatQuantifierEliminator(QuantifierEliminator, IdentityDagWalker):
+
+        def __init__(self, environment, algorithm='fm'):
+            """Algorithm can be either 'fm' (for Fourier-Motzkin) or 'lw' (for
+               Loos-Weisspfenning)"""
+
+            IdentityDagWalker.__init__(self, env=environment)
+
+            self.functions[op.SYMBOL] = self.walk_identity
+            self.functions[op.REAL_CONSTANT] = self.walk_identity
+            self.functions[op.BOOL_CONSTANT] = self.walk_identity
+            self.functions[op.INT_CONSTANT] = self.walk_identity
+
+
+            assert algorithm in ['fm', 'lw']
+            self.algorithm = algorithm
+
+            self.config = mathsat.msat_create_default_config("QF_LRA")
+            self.msat_env = mathsat.msat_create_env(self.config)
+            self.converter = MSatConverter(environment, self.msat_env)
+
+
+        def eliminate_quantifiers(self, formula):
+            """
+            Returns a quantifier-free equivalent formula of the given
+            formula
+            """
+            return self.walk(formula)
+
+
+        def exist_elim(self, variables, formula, alg='fm'):
+            theory = self.env.theoryo.get_theory(formula)
+            logic = Logic(name="Detected Logic", description="",
+                          quantifier_free=False, theory=theory)
+            logic = get_closer_pysmt_logic(logic)
+            if not logic <= LRA:
+                raise NotImplementedError("MathSAT qunatifier elimination only"\
+                                          " supports LRA (detected logic " \
+                                          "is: %s)" % str(logic))
+
+            fterm = self.converter.convert(formula)
+            tvars = [self.converter.convert(x) for x in variables]
+
+            assert alg in ['fm', 'lw']
```

Comment:
Is the function exist_elim designed to be called directly?
If no, then the "alg" argument is not needed, since we have self.alg .

Moreover, I would convert these 3 lines into a map.